firebaseai/LiveAudioExample/ViewModels/LiveViewModel.swift

```diff
@@ -0,0 +1,295 @@
+      let time = goingAwayNotice.timeLeft?.description ?? "soon"
+      logger.warning("Going away in: \(time)")
+    }
+  }
+
+  private func processServerContent(_ content: LiveServerContent) async {
+    if let message = content.modelTurn {
+      await processAudioMessages(message)
+    }
+
+    if content.isTurnComplete {
+      // add a space, so the next time a transcript comes in, it's not squished with the previous one
+      transcriptViewModel.appendTranscript(" ")
+    }
+
+    if content.wasInterrupted {
+      logger.warning("Model was interrupted")
+      await audioController?.interrupt()
+      transcriptViewModel.clearPending()
+      // adds an em dash to indiciate that the model was cutoff
+      transcriptViewModel.appendTranscript("— ")
+    } else if let transcript = content.outputAudioTranscription?.text {
+      appendAudioTranscript(transcript)
+    }
+  }
+
+  private func processAudioMessages(_ content: ModelContent) async {
+    for part in content.parts {
+      if let part = part as? InlineDataPart {
+        if part.mimeType.starts(with: "audio/pcm") {
+          await audioController?.playAudio(audio: part.data)
+        } else {
+          logger.warning("Received non audio inline data part: \(part.mimeType)")
+        }
+      }
+    }
+  }
+
+  private func processFunctionCalls(functionCalls: [FunctionCallPart]) async {
+    let responses = functionCalls.map { functionCall in
+      switch functionCall.name {
+      case "changeBackgroundColor":
+        return changeBackgroundColor(args: functionCall.args, id: functionCall.functionId)
+      case "clearBackgroundColor":
+        return clearBackgroundColor(id: functionCall.functionId)
+      default:
+        logger.debug("Function call: \(String(describing: functionCall))")
+        fatalError("Unknown function named \"\(functionCall.name)\".")
+      }
+    }
+
+    await liveSession?.sendFunctionResponses(responses)
+  }
+
+  private func appendAudioTranscript(_ transcript: String) {
+    transcriptViewModel.appendTranscript(transcript)
+  }
+
+  private func changeBackgroundColor(args: JSONObject, id: String?) -> FunctionResponsePart {
+    guard case let .string(color) = args["color"] else {
+      logger.debug("Function arguments: \(String(describing: args))")
+      fatalError("Missing `color` parameter.")
+    }
+
+    withAnimation {
+      backgroundColor = Color(hex: color)
+    }
+
+    if backgroundColor == nil {
+      logger.warning("The model sent us an invalid hex color: \(color)")
+    }
+
+    return FunctionResponsePart(
+      name: "changeBackgroundColor",
+      response: JSONObject(),
+      functionId: id
+    )
+  }
+
+  private func clearBackgroundColor(id: String?) -> FunctionResponsePart {
+    withAnimation {
+      backgroundColor = nil
+    }
+
+    return FunctionResponsePart(
+      name: "clearBackgroundColor",
+      response: JSONObject(),
+      functionId: id
+    )
+  }
+}
+
+extension Color {
```

Comment:
This should probably go into a separate `Color+Hex.swift` file, so others are aware and can use it as well.